include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,169 @@
+  UNIFEX_NO_UNIQUE_ADDRESS Predecessor pred_;
+  UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+  UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template <template <typename...> class Tuple>
+  struct indexed_for_result {
+   public:
+    template <typename... Args>
+    using apply = Tuple<Args...>;
+  };
+
+  template <typename... Args>
+  using is_overload_noexcept = std::bool_constant<noexcept(
+      std::invoke(std::declval<Func>(), std::declval<Args>()...))>;
+
+  template <template <typename...> class Variant>
+  struct calculate_errors {
+   public:
+    template <typename... Errors>
+    using apply = std::conditional_t<
+        Predecessor::
+            template value_types<std::conjunction, is_overload_noexcept>::value,
+        Variant<Errors...>,
+        deduplicate_t<Variant<Errors..., std::exception_ptr>>>;
+  };
+
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = deduplicate_t<typename Predecessor::template value_types<
+      Variant,
+      indexed_for_result<Tuple>::template apply>>;
+
+  template <template <typename...> class Variant>
+  using error_types = typename Predecessor::template error_types<
+      calculate_errors<Variant>::template apply>;
+
+  friend constexpr auto tag_invoke(
+      tag_t<blocking>,
+      const indexed_for_sender& sender) {
+    return blocking(sender.pred_);
+  }
+
+  template <typename Receiver>
+  struct indexed_for_receiver {
+    UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+    UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+    UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+    // sequenced_policy version supports forward range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::sequenced_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      for(auto idx : range) {
+        std::invoke((Func &&) func, idx, values...);
+      }
+    }
+
+    // parallel_policy version requires random access range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::parallel_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      auto start = range.begin();
+      for(auto idx = 0; idx < range.size(); ++idx) {
+        std::invoke((Func &&) func, start[idx], values...);
+      }
+    }
+
+    template <typename... Values>
+    void set_value(Values&&... values) && noexcept {
+      if constexpr (noexcept(std::invoke(
+                        (Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+        apply_func_with_policy(policy_, (Range&&) range_, (Func &&) func_, values...);
```

Comment:
What should the behaviour be here if you pass a policy of `execution::par` but pass a range that is an input or forward range? Should it fall back to the sequential implementation?

The existing parallel algorithms often only require forward-iterators rather than random-access iterators.

Should this also be tag-dispatching based on the `iterator_category`?

Reply:
I think it should fail. We can't fully sanity check applications of func to see if they are par-safe, but requiring random access is a minimal check on the iterator. I was planning to change the text in P1897 to just require random_access for non-seq policies, forward for seq.

I moved the policy before the range because I think it is a declaration of the correctness of access to both the range and the function. If you do that and pass a forward range then that is immediately undefined behaviour by the terms of the declaration the programmer is making, but it's one we can check.

Reply:
Could you relax it further to 'input range' for seq policy, at least for single-pass algorithms like this?
That would allow using things like generators as the range.

Reply:
Yes, I think so.